torch/distributed/_shard/checkpoint/utils.py

```diff
@@ -0,0 +1,193 @@
+                object_list=gather_objs,
+                obj=object,
+                group=self.group
+            )
+        else:
+            gather_objs = [object]
+        return gather_objs
+
+    def scatter_object(self, object_list: List[Any]) -> Any:
+        if self.use_dist:
+            gather_result = [None]
+            dist.scatter_object_list(
+                scatter_object_output_list=gather_result,
+                scatter_object_input_list=object_list if self.is_coordinator else None,
+                src=self.coordinator_rank,
+                group=self.group
+            )
+
+            local_reply = gather_result[0]
+        else:
+            local_reply = object_list[0]
+        return local_reply
+
+    def reduce_scatter(self,
+        step: str,
+        map_fun: Callable[[], Any],
+        reduce_fun: Callable[[List[Any]], List[Any]]
+    ) -> Any:
+        try:
+            local_data = map_fun()
+        except BaseException as e:
+            local_data = e
+
+        all_data = self.gather_object(local_data)
+        all_results = None
+        if self.is_coordinator:
+            node_failures = {i: err for i, err in enumerate(all_data) if isinstance(err, BaseException)}
+
+            if len(node_failures) == 0:
+                try:
+                    all_results = reduce_fun(all_data)
+                except BaseException as e:
+                    node_failures[self.rank] = e
+
+            if len(node_failures) > 0:
+                all_results = [CheckpointException(step, node_failures)] * self.get_world_size()
+
+        result = self.scatter_object(all_results)
+        if isinstance(result, CheckpointException):
+            raise result
+        return result
+
+    def all_reduce(self,
+        step: str,
+        map_cb: Callable[[], Any],
+        reduce_cb: Callable[[List[Any]], Any]
+    ) -> Tuple[Any, Any]:
+        try:
+            local_data = map_cb()
+        except BaseException as e:
+            local_data = e
+
+        all_data = self.gather_object(local_data)
+        result = None
+        if self.is_coordinator:
+            node_failures = {i: err for i, err in enumerate(all_data) if isinstance(err, BaseException)}
+
+            if len(node_failures) == 0:
+                try:
+                    result = reduce_cb(all_data)
+                except BaseException as e:
+                    node_failures[self.rank] = e
+
+            if len(node_failures) > 0:
+                result = CheckpointException(step, node_failures)
+
+        result = self.broadcast_object(result)
+        if isinstance(result, CheckpointException):
+            raise result
+        return result
+
+    def all_gather(self,
+        step: str,
+        map_fun: Callable[[], Any],
+    ) -> Tuple[Any, Any]:
+        try:
+            result = map_fun()
+        except BaseException as e:
+            result = e
+
+        all_results = self.all_gather_object(result)
```

Comment:
is there a reason we use `all_gather_object` instead of `all_gather`?

Reply:
yes, the `_object` variants handle dist not initialized.